### DISCUSSION TITLE
How to prevent starting two instances of rust-analyzer with `mason` (LunarVim)?

### DISCUSSION THREAD

**Discussion by Logarithmus** [mrcjkb/rustaceanvim#174] [Q&A] [answered]
Hello. I used to use rust-tools.nvim with LunarVim. After its deprecation, I switched to rustaceanvim. Due to the fact this plugin wants to manage rust-analyzer startup on its own, I end up having 2 rust-analyzers running: one launched by rustaceanvim and another by either mason or nvim-lspconfig (I'm still not sure who of them is responsible for starting language servers).

How can I use this plugin together with LunarVim?
Is there a possibility to disable rustaceanvim's rust-analyzer autostart, and make rustaceanvim to connect to already launched rust-analyzer instead?

**Comment by mrcjkb** [comment]
Hey :wave: 

Unfortunately, I can't provide official support for all Neovim distros out there.
Your question seems related to this one: https://github.com/mrcjkb/rustaceanvim/discussions/94

As you're using LunarVim, the more general solution should work for you.

**Reply by Logarithmus** [reply, marked as answer]
Finally, after wasting several hours on this, as a person who isn't familiar enough with lua or LunarVim internals, I came up with the solution! My problem was finding out how to apply LunarVim's LSP keymaps to `rustaceanvim`. Huge thanks to [this person's config](https://github.com/aeet/lvim_conf/blob/6bb85016f19a19e2340ce449fea472a70a5c6376/lua/user/rust_tools.lua#L36)!
1. Add this line into `~/.config/lvim/config.lua`: 
```lua
vim.list_extend(lvim.lsp.automatic_configuration.skipped_servers, { "rust_analyzer" })
```
2. Your `rustaceanvim` entry in `lvim.plugins` should look like this:
```lua
{
  "mrcjkb/rustaceanvim",
  version = "^3",
  ft = { "rust" },
  config = function()
    vim.g.rustaceanvim = {
      server = {
        on_attach = require("lvim.lsp").common_on_attach
      },
    }
  end,
}
```

**Reply by TheBearodactyl** [reply]
I tried this and it still spawns a second instance on my install. Any way to fix that?

**Reply by mrcjkb** [reply]
@TheBearodactyl you would have to provide more details about your setup.

See also [this answer to another related discussion](https://github.com/mrcjkb/rustaceanvim/discussions/94#discussioncomment-7813716) that provides a more general solution.

**Reply by utensil** [reply]
Following this leads to an error on rust file open:

```
nvim-navic: Failed to attach to rust_analyzer for current buffer. Already attached to rust-analyzer
```

**Reply by mrcjkb** [reply]
@utensil You are using `lspconfig.rust_analyzer`, which conflict's with rustaceanvim.

**Reply by GaussianWonder** [reply]
Also execute `:LvimCacheReset` after config updates in order for `vim.list_extend(lvim.lsp.automatic_configuration.skipped_servers, { "rust_analyzer" })` to take effect.

**Reply by utensil** [reply]
To prevent LunarVim from spinning up its rust_analyzer, I used the following in my config:

```lua
-- to prevent colision with rusteceanvim
-- https://github.com/mrcjkb/rustaceanvim/discussions/174#discussioncomment-8193827
vim.list_extend(lvim.lsp.automatic_configuration.skipped_servers, { "rust_analyzer" })

-- In order for the above to work, one must execute `:LvimCacheReset` manually'
-- or schedule the following
vim.schedule(function()
    vim.cmd "LvimCacheReset"
end)
```

So I don't have to reset cache manually after finding out two rust analyzers eating all CPU cores.

**Comment by ArcusCerebellumus** [comment]
@TheBearodactyl 
> I tried this and it still spawns a second instance on my install. Any way to fix that?

Just leaving it here for posterity...

I don't use LunarVim, but I had the same happen to me just the other day. Came up with a bunch of solutions most simple of which is just deleting rust_analizer from Mason. However, If you want to keep it, and/or want more flexibility - you can add a condition that prevents certain servers from ever being configured (that's what I did):
``` lua
---@type MasonLspconfigSettings
require('mason-lspconfig').setup({

    ---@diagnostic disable-next-line
    handlers = { function(server_name)
    
        -- a quick hack to check the list (...if you need a LIST)
        local contains = function(table, value)
            for index = 1, #table do
                if table[index] == value then
                    return true
                end
            end
            return false
        end

        -- exclusion list
        local exclude = {
            -- don't want rust, as I'm using Rustacianvim for it
            'rust_analyzer',
        }
...
```
Next, you add a condition:
``` lua
...
        if contains(exclude, server_name) then
            print('NOPE!')
...
```
A simpler solution, if you only want to override for rust - throw out the top part and just:
``` lua
...
        if server_name == 'rust_analizer' then
            print('NOPE!')
...
```
the rest is just your typical stuff that you'd do in mason-lspconfig:

``` lua
...           
        else
            -- get server settings IF they are specified in server_settings above
            local server = server_settings[server_name] or {}
            -- get default capabilities and override/extend them IF they are in server_settings
            server.capabilities = vim.tbl_deep_extend(
                'force',
                {},
                capabilities,
                server.capabilities or {}
            )
            -- run setup
            require('lspconfig')[server_name].setup(server)
        end
    end, },
})
```
There might already be a special function or a config option for that in mason-lspconfig, but I was too lazy to check and just made this monstrosity. Hope it helps.
